Added Rest Client Module to benchmark Restful Web Services

shivam-maharshi opened the pull request:
Please review. Touched only the absolutely necessary files required. CheckStyle formatting is followed, unit tests created, build passes.

shivam-maharshi commented:
@busbey Please give feedback when you get some time. I have incorporated all your old comments.

shivam-maharshi commented:
@risdenk @kruthar In case you get sometime to take a look at this. Please do and provide feedback.

kruthar commented:
Hi @shivam-maharshi - I have been looking at this PR on and off. For me it is hard to review for a couple reasons:

1. The PR description doesn't give any detail into what this PR is trying to accomplish, I'm hesitant to approve this because I don't know really what it is for or what exactly it does, or any of the motivation behind it, why it was necessary to create a new Workload
2. Following the documentation theme there isn't a README.md which would tell me a little more about the Rest driver, what it is for, how to use it, the properties, etc.

I started looking into the RestWorkload.java, it is very different from CoreWorkload.java and so is not easy to view differences between the two with a diff. This reinforces the need for more documentation around what the motivations for this PR are and the benefits it brings.

Some specific things that are confusing to me:

1. I am not exactly sure what this trace stuff is for. You have separated out trace properties for each operation, but then in the workload file examples you have just given the same file name for each operation. what is the point there?
2. It seems strange to me that at the RestWorkload.java level you have removed the scan operation.
3. It looks like the key and value generating paradigms have changed, especially since buildDeterministicValue doesn't exist in RestWorkload.java

Sorry for the negative feedback but some things I would really like to see are more documentation and conversation about the work you've done, at this point there is too much change to really tell everything that is going on. Along with that it would be great to discuss if the changes you have made in RestWorkload.java could actually be integrated through additions to CoreWorkload.java and the bulk of the changes in the `rest` binding you created.

shivam-maharshi commented:
Hi @kruthar, thanks a lot for providing a detailed explanation. My aim is to eventually get this PR accepted. I am open to making any changes that are required to make it possible. Any feedbacks that will help me move in this direction are more than welcome! Here are the answers to some of your concerns.

1. I have added a detailed README.md for this driver. It includes description about what this driver is, what is the use case, how to use it and the description of the properties to be used with examples. This was an important point, thanks for pointing it out.

2. This main aim of this PR is to add the support of benchmarking RESTFul webservices using YCSB. Right now YCSB can only benchmark databases, however it has the potential to be an excellent tool for benchmarking webservices mainly because of three reasons. Firstly, it is highly configurable and customizable. Secondly, it is capable of producing huge amount of workload without boging the client machine down, something needed for cloud services. Thirdly, since it is written on java, it is platform independent. I have performed a detailed study into the available free opensource tools most suited for benchmarking cloud webservices, and found that there is a lack of such a tool (My MS thesis is on a related topic). Hence I finally decided to add this module to YCSB since it catered to my need and I feel that many developers would be thrilled to use YCSB it. They can also easily tweak it to fit thier own usecase.

3. Regarding RestWorkload, below are the explanations:
a. Trace is a file that holds the path of URL resources that are to be used for making HTTP requests to benchmark your web service. For example, if your webservice exists at http://{host}:{port}/{endpoint}, and you want to benchmark the performance of READS on this webservice then the url.trace.read file might look like this:

http://{host}:{port}/{endpoint}/resource_1
http://{host}:{port}/{endpoint}/resource_2
http://{host}:{port}/{endpoint}/resource_3

The RestWorkload will then pickup these URL resources and put them in a map. According to the requestdistribution type, the URLs will be picked from this map and passed to RestClient binding to make HTTP GET requests on these URLs. Similar logic will also apply to INSERT, DELETE and UPDATE methods.

b. In real life the URL access pattern is different for different operations. For example, HTTP GET will have a different URL access pattern than HTTP POST for a webservice. Hence there must be different trace files for each such operation. However for simplicity, we might want to use similar trace file. This is the case here. 

c. The support for SCAN operations are important to benchmark from a database perspective where sometimes a collection of entries are fetched or updated for some application layer logic. This is a real life use case for databases. However for webservices this is highly unlikely use case. Hence the support for this operation is not needed. However it can always be added in the future if deemed worth the time and efforts.

d. In RestWorkload, the keys generated are simply used to pick a URL from the trace maps, loaded at initialization of the class. Using this technique we are able to decide which URL should the next request be made so as to produce a URL access pattern similar to real life. The buildDeterministicValue is hence not required in this use case.

I totally understand your concerns and I am more than happy to discuss with you the reasoning behind this binding and the design opted. We can have a conversation on phone/email/skype or any other convenient option as you prefer. If you or anyone else have better suggestions regarding design and we would need to restructure the code, I would be more than happy to do that myself as well. Thanks again for your response. Hoping to hear back soon again.

kruthar commented:
@shivam-maharshi - thanks for the background info and context, it makes a lot more sense now why this type of workload needs to be so different from the traditional CoreWorkload. I think you have also worked in enough flexibility to work with many potential REST layouts by using the trace files, a little more work for the user up front, but it means they can be sure it will hit all the endpoints they want it to and they dont seem to have to heterogenous.

@busbey - there are several things that need to be updated in this PR before being merged in, but do you have any problems with trying to get this one in for 0.10.0 as an Untested binding? I would put it in as untested to start with because of the fact that it introduces a new Workload type, and I don't have a REST services on hand to test this thing.

kruthar commented on rest/README.md:
```diff
@@ -0,0 +1,174 @@
+The rest module will pick up URLs from the above file according to
+the `requestdistribution` property (default is zipfian) mentioned in
+the rest_workload. In the example above we assume that the property `url.prefix` 
+(see below for property description) is set to ``. If url.prefix
```
nit - there are two backticks here that are throwing off formatting.

shivam-maharshi replied:
Fixed it by using empty keyword instead of ``.

kruthar commented on rest/pom.xml:
```diff
@@ -0,0 +1,80 @@
+  <parent>
+    <groupId>com.yahoo.ycsb</groupId>
+    <artifactId>binding-parent</artifactId>
+    <version>0.10.0-SNAPSHOT</version>
```
A release has just been started, update the snapshot version to 0.11.0-SNAPSHOT

shivam-maharshi replied:
Updated snapshot to 0.11.0-SNAPSHOT

kruthar commented:
Can you rebase this PR to pick up the new ycsb shell scripts, then add a line for this rest binding in bin/bindings.properties. The python tool at bin/ycsb will be deprecated out at some point, but you might as well update that also with a line for your binding.

kruthar commented:
Would also be greatly appreciated if you could rebase your PR done to just a handful of meaningful commits. Please prefix each commit with `[rest]` as that is the subject matter.

shivam-maharshi commented:
@kruthar Rebased the PR and updated bin/binding.properties and python bin/ycsb to include rest module. Also squashed all the different commits into a single commit and used [rest] in the message log. However I see all the commits by other developers (made in the last month) into this PR as well. I think I made a mistake somewhere rebasing my branch. Not sure how to fix this. Would this PR work or should I close this and create a new PR? Please let me know and I'll do the same.

kruthar commented:
@busbey, @risdenk - looks like this is failing with the jdk8 test due to the sigar issue addressed in PR #776. I don't see any other recent PR's failing so thinking this is kinda weird. Thoughts? Is #776 required to do jdk8 builds now?

risdenk commented:
@kruthar The sigar thing is just a warning. It doesn't actually fail anything. The real issue in the test is `Killed` at the end after

```
Connected to cluster: Test Cluster
Datacenter: datacenter1; Host: localhost/127.0.0.1; Rack: rack1
Connected to cluster: Test Cluster
Datacenter: datacenter1; Host: localhost/127.0.0.1; Rack: rack1
Connected to cluster: Test Cluster
Datacenter: datacenter1; Host: localhost/127.0.0.1; Rack: rack1
```

The real issue is the JVM crashed or hung or something. Maybe out of memory or something from Travis?
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test (default-test) on project cassandra2-binding: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test failed: The forked VM terminated without saying properly goodbye. VM crash or System.exit called ? -> [Help 1]
```

busbey commented:
> @busbey - there are several things that need to be updated in this PR before being merged in, but do you have any problems with trying to get this one in for 0.10.0 as an Untested binding? I would put it in as untested to start with because of the fact that it introduces a new Workload type, and I don't have a REST services on hand to test this thing.

I'd rather take the time to get it right before merging. 0.11.0 will come out in short order after 0.10.0.

jixuan1989 commented:
Sorry, I am not sure that sigar is only suitable for jdk8. And if we do not have sigar files, there will be exception information in the console: no libsigar-universal64-macosx.dylib in java.library.path. 

But as the above said, maybe Cassandra can still startup (I am not sure).

shivam-maharshi commented:
@kruthar @busbey Please share your feedbacks about what all needs to be updated in the PR. I will be more than happy to work with you guys on this.

kruthar commented on bin/ycsb:
```diff
@@ -83,6 +83,7 @@ DATABASES = {
     "orientdb"     : "com.yahoo.ycsb.db.OrientDBClient",
     "rados"        : "com.yahoo.ycsb.db.RadosClient",
     "redis"        : "com.yahoo.ycsb.db.RedisClient",
+	"rest"         : "com.yahoo.ycsb.webservice.rest.RestClient",
```
nit - line this up

kruthar commented on rest/pom.xml:
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (c) 2011 YCSB project, 2014 - 2016 YCSB contributors. All 
```
This is a new file it can just have 2016 YCSB Contributors

kruthar commented on rest/src/test/resources/workload_rest:
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2016 Yahoo! Inc. All rights reserved.                                                                                                                             
```
Is it necessary to have this copy of the file? The copy in workloads/ will be distributed.

shivam-maharshi replied:
Its not necessary but good to have. Since we've already mentioned a detailed README.md, this can be removed without implications.

risdenk replied:
Looks like you said this would be removed and it hasn't been?

shivam-maharshi replied:
This is needed now for the Integration Test cases. Hence hasn't been removed.

risdenk replied:
ok thanks for the explanation.

kruthar commented:
license edit in 
- bin/ycsb
- distribution/pom.xml

Does anyone else have a REST service just laying around to test this with?

shivam-maharshi commented:
@kruthar Incorporated feedbacks. For starters you can just test read operations using 5 Google URLs in the trace file like this:

url.prefix=https://www.google.com/#q=

**TraceFile:**
study
education
aws
cloud+computing
climate+change

shivam-maharshi commented:
@kruthar Please let me know if you need some help setting a dummy Rest service. We can work together on verifying this anytime you are free.

busbey commented:
as an alternative for unblocking testing, what if we add an integration test that stands up an embedded REST service and then runs against it?

shivam-maharshi commented:
@busbey We are using Wiremock for imitating Rest services (here: https://github.com/brianfrankcooper/YCSB/pull/756/files#diff-237d7ac48706b45c30b7c2ce9cec005eR97) in the RestClientTests, which take care of the HTTP handling part. It can also be used for integration test cases. On a side note I can add unit tests for RestWorkload to increase test coverage, thereby making sure that this module works as integration test cases might take sometime.

busbey commented:
I saw the use of Wiremock for the unit tests, I was hoping for something more thorough for an integration test. Mocks are great for verifying well-encapsulated assumptions about the client, but tend to miss out on the sorts of edge cases that come from talking to an actual system. I was hoping for something like an embedded tomcat demo or the like.

NBD if there isn't something on hand. We'll just need to wait for somebody to have time to test things out. (that needn't be @kruthar btw, or even any of the existing project moderators. we accept reviews from anybody.)

shivam-maharshi commented:
@busbey @kruthar I've added integration test cases using embedded Tomcat 8. These tests start the embedded server (once at class loading) which exposes a mock RESTFul service used for benchmarking. Benchmarking is invoked using Client class from the core module to generate load using the RestWorkload. Everything works as expected. Please review the changes and let me know.

shivam-maharshi commented:
@channel I've confirmed the working of the rest-binding both manually as well as using integration test cases. Please let me know if it needs more changes to get accepted.

shivam-maharshi commented:
@risdenk @jasontedor Can you please take a look at this pull request as the other moderators are busy.

risdenk commented:
@shivam-maharshi I'll try to look on Friday.

shivam-maharshi commented:
@risdenk thanks. Even though integration tests runs fine, please let me know if you need some help testing this on a real RESTFul service.

risdenk commented on rest/README.md:
```diff
@@ -0,0 +1,181 @@
+
+    git clone git://github.com/brianfrankcooper/YCSB.git
+    cd YCSB
+    mvn -pl com.yahoo.ycsb:solr-binding -am clean package
```
This refers to solr-binding instead of rest?

shivam-maharshi replied:
Corrected

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/IntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+  public final ExpectedSystemExit exit = ExpectedSystemExit.none();
+
+  private static int port = 8080;
+  private static Tomcat tomcat = new Tomcat();
```
Should this be in the init()?

shivam-maharshi replied:
Moved.

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/IntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+
+  private static int port = 8080;
+  private static Tomcat tomcat = new Tomcat();
+  private static final String BASE_DIR = System.getProperty("user.dir");
```
You should be able to get the paths to these files with Java resources. Some psuedo code:

```
this.class.getClassLoader().getResource()
```

shivam-maharshi replied:
Thanks! Code Snippet - `IntegrationTest.class.getClassLoader().getResource("error_trace.txt").getPath();`

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/IntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+  @BeforeClass
+  public static void init() throws ServletException, LifecycleException, FileNotFoundException, IOException,
+      DBException, InterruptedException {
+    String webappDirLocation = "WebContent/";
```
Can WebContent be moved to test/resources?

shivam-maharshi replied:
Moved

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/RestClientTest.java:
```diff
@@ -0,0 +1,215 @@
+ * demonstrated.
+ */
+ // Test cases pass for JDK8 but fail for JDK7 due to WireMock.
+ @Ignore 
```
This whole test class is ignored?

shivam-maharshi replied:
Wiremock (The framework used to mock rest services for RestClientTest) requires JDK8 support. Hence this class is ignored as the tests fail on JDK7 but pass on JDK8.

risdenk replied:
Should be able to ignore tests for a specific JDK. See https://github.com/brianfrankcooper/YCSB/blob/master/cassandra/pom.xml#L67

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/RestClientTest.java:
```diff
@@ -0,0 +1,215 @@
+  @BeforeClass
+  public static void init() throws FileNotFoundException, IOException, DBException {
+    Properties props = new Properties();
+    props.load(new FileReader("src/test/resources/workload_rest"));
```
Again should be able to use resources loading instead of a full path?

shivam-maharshi replied:
Done. Code Snippet - 
`props.load(new FileReader(RestClientTest.class.getClassLoader().getResource("workload_rest").getPath()));`

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/RestWorkload.java:
```diff
@@ -0,0 +1,300 @@
+/**
```
I think more time needs to be spent removing this RestWorkload file. I think it overlaps with CoreWorkload.

The format of CoreWorkload and RestWorkload of how it documents the properties and defines variables is different I think. Maybe try to make them more similar?

shivam-maharshi replied:
I agree with you that it would be nice to have RestWorkload merged into CoreWorkload but there are some concerns/points.
1. In order to benchmark REST-full web services we must already have URL endpoints to benchmark. This is why we have URL maps in RestWorkload. However the way core workload is written currently, it generates keys to benchmark instead of reading them from a collection of predefined keys. Which definitely makes sense for DBs but not for web services.
2. Considering the point above, if we still merge RestWorkload into CoreWorkload, it would require us to distinguish the key generation logic on the basis of some config value which would make the CoreWorkload code ugly and couple it with logic specific with a particular binding.
3. Delete Operations are not supported in CoreWorload, they need special handling for WebServices to benchmark.
4. Since CoreWorkload is the central component, updating it will require an exhaustive testing for all the other bindings. This is precisely what I wanted to avoid.
5. To minimize code duplication RestWorkload extends CoreWorkload. Hence most future updates in CoreWorkload will be propagated to RestWorkload automatically thereby reducing extra code efforts.

I understand the urge to have a single workload but given the different nature of WebServices there are some inherent differences from databases which makes this task a bit non-trivial. I am open to further improve this design in the future but for now, if we can get this pull request incorporated in the master then I will create another pull request with the improved design later.

risdenk replied:
Fair enough we can always improve it later.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/RestWorkload.java:
```diff
@@ -0,0 +1,300 @@
+  private static final String UPDATE_ZIPFIAN_CONSTANT_DEAFULT = "0.99";
+  private static final String UPDATE_RECORD_COUNT_PROPERTY = "updaterecordcount";
+
+  private static Map<Integer, String> readUrlMap;
```
Should these all be static? Can these be accessed by different threads?

shivam-maharshi replied:
Good point! They will be accessed by different threads but only for reading. But yes it makes sense to not keep them static as a single YCSB instance can be used for multiple benchmarking with different traces. Change incorporated.

risdenk commented on rest/pom.xml:
```diff
@@ -0,0 +1,144 @@
+    <dependency>
+      <groupId>org.apache.httpcomponents</groupId>
+      <artifactId>httpclient</artifactId>
+      <version>4.5.1</version>
```
For consistency, can you pull all the versions into the properties above like you did for tomcat and jersey?

risdenk commented:
@shivam-maharshi - Can you ping this thread after you update? Thanks!

shivam-maharshi commented:
@risdenk I have incorporated all your feedbacks except for one (Merging RestWorkload into CoreWorkload.). Please take a look.

shivam-maharshi commented:
@risdenk @busbey Please let me know if we need further modifications in this PR. Thanks!

risdenk commented on rest/pom.xml:
```diff
@@ -25,6 +25,12 @@
+    <junit.version>4.12</junit.version>
+    <system-rules.version>1.16.0</system-rules.version>
+    <wiremock.version>1.57</wiremock.version>
+    <skipTests>true</skipTests>
```
Any reason skipTests is here?

shivam-maharshi replied:
Removed. An error during commit. Build was passing.

risdenk commented on rest/pom.xml:
```diff
@@ -0,0 +1,150 @@
+    </dependency>
+  </dependencies>
+  
+  <build>
```
Does this plugin need to be here?

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/RestWorkload.java:
```diff
@@ -0,0 +1,300 @@
+      keychooser = new ZipfianGenerator(recordCount, zipfContant);
+    } else if (requestDistrib.compareTo("latest") == 0) {
+      // TODO: To be implemented.
+      System.err.println("Latest request distribution is not supported");
```
This should exit instead of returning null?

risdenk replied:
The error is printed without throwing a WorkloadException

shivam-maharshi replied:
Handled - `throw new WorkloadException("Latest request distribution is not supported for RestWorkload.");`

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/RestWorkload.java:
```diff
@@ -0,0 +1,300 @@
+  private static NumberGenerator getKeyChooser(String requestDistrib, int recordCount, double zipfContant,
+      Properties p) throws WorkloadException {
+    NumberGenerator keychooser = null;
+    if (requestDistrib.compareTo("exponential") == 0) {
```
each of these `.compareTo` should just be `.equals()`?

risdenk replied:
May even be cleaner as a switch case

shivam-maharshi replied:
I had the same opinion but kept it consistent with the older CoreWorkload code. Updated to switch statements now.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/RestWorkload.java:
```diff
@@ -0,0 +1,300 @@
+   */
+  private static synchronized Map<Integer, String> getTrace(String filePath, int recordCount)
+      throws WorkloadException {
+    Map<Integer, String> urlMap = new ConcurrentHashMap<Integer, String>();
```
Since this is a synchronized method does this need to be ConcurrentHashMap?

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,356 @@
+    int responseCode;
+    try {
+      responseCode = httpExecute(new HttpPut(urlPrefix + endpoint), values.get("data").toString());
+    } catch (Exception e) {
```
This is catching all exceptions and not actually printing what happened? Would be nice to log the actual exception?

risdenk replied:
Might be able to just add the logging to the handleExceptions method

shivam-maharshi replied:
Done

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,356 @@
+        this.timedout.setIsSatisfied(true);
+      } catch (InterruptedException e) { 
+        // Do nothing. Kept just to remove Check-style build error.
+        byte a = 0;
```
Hmmm wonder if there is a better way to do this.

shivam-maharshi replied:
By updating checkstyle.xml with the below snippet a block now can have some text as opposed to just code to be considered valid:
`<module name="EmptyBlock">
     <property name="option" value="text"/>
 </module>`

risdenk commented on rest/src/test/java/com/yahoo/ycsb/webservice/rest/IntegrationTest.java:
```diff
@@ -0,0 +1,245 @@
+  private static final String WORKLOAD_FILEPATH =  IntegrationTest.class.getClassLoader().getResource("workload_rest").getPath();
+  private static final String TRACE_FILEPATH = IntegrationTest.class.getClassLoader().getResource("trace.txt").getPath();
+  private static final String ERROR_TRACE_FILEPATH = IntegrationTest.class.getClassLoader().getResource("error_trace.txt").getPath();
+  private static final String RESULTS_FILEPATH = IntegrationTest.class.getClassLoader().getResource("").getPath() + "results.txt";
```
the `getResource("")` calls shouldn't be to `""` they should be to the proper file or directory.

shivam-maharshi replied:
The file results doesn't exist hence we can't call `getResource("results.txt");`. Since we need the folder path to /src/test/resources even `getResource("resources")` doesn't work. Hence I've used a period - '.' signifying the current target directory.

shivam-maharshi commented:
@risdenk I have incorporated all your feedbacks. The build is failing due to some Cassandra test cases. Can you please take a look. 

Additionally, someone reached out to me asking for REST Web Services support in YCSB (https://github.com/shivam-maharshi/YCSB4WebServices/issues/1). It would be great if we can get this PR accepted soon so people can make use of it.

risdenk commented:
@shivam-maharshi - kicked off the test again and looks like it failed related to the rest piece:

```
Failed tests:   insert_200(com.yahoo.ycsb.webservice.rest.RestClientTest): expected:<Status [name=OK, description=The operation completed successfully.]> but was:<Status [name=ERROR, description=The operation failed.]>

Tests run: 26, Failures: 1, Errors: 0, Skipped: 0
```

shivam-maharshi commented:
@risdenk This error occurs sometime in JDK8 and sometime in JDK7 intermittently. It because some request in the RestClientTest is taking more than 10 seconds to execute and hence timer thread kills it and throws an Error Status. Right now the timeout is 10 seconds which I thought would be enough. What do you propose we do to resolve this? I am going to increase the timeout to 20 seconds and see if that passes. However in long run this is bad to build runtime. If you have any suggestions to handle this elegantly, please let me know.

shivam-maharshi commented:
@risdenk All checks pass now. Please take a look and lemme know.

risdenk commented:
@shivam-maharshi Travis CI for whatever reason tends to test really slowly. This seems to cause a lot of issues with Accumulo as well. Raising the timeout shouldn't hurt right? If it the test finishes then the timeout has no effect.

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,367 @@
+
+  private int handleExceptions(Exception e, String url, String method) {
+    if (logEnabled) {
+      System.out.println(new StringBuilder(method).append(" Request: ").append(url).append(" | ")
```
The `System.out` in a few places should be `System.err` to make sure that messages go to stderr. I think there are a few examples of this in other bindings.

shivam-maharshi replied:
@risdenk Let me quickly fix the logging and squash all the commit with [rest] today itself. Will notify you with the next commit. Thanks!

shivam-maharshi commented:
@risdenk Incorporated feedbacks and squashed all the commits into one. Please take a look and lemme know.

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,370 @@
+      responseCode = handleExceptions(e, urlPrefix + endpoint, HttpMethod.POST);
+    }
+    if (logEnabled) {
+      System.out.println(new StringBuilder("POST Request: ").append(urlPrefix).append(endpoint)
```
`System.out` instead of `System.err`

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,370 @@
+      responseCode = handleExceptions(e, urlPrefix + endpoint, HttpMethod.DELETE);
+    }
+    if (logEnabled) {
+      System.out.println(new StringBuilder("DELETE Request: ").append(urlPrefix).append(endpoint)
```
`System.out` instead of `System.err`

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,370 @@
+      responseCode = handleExceptions(e, urlPrefix + endpoint, HttpMethod.GET);
+    }
+    if (logEnabled) {
+      System.out.println(new StringBuilder("GET Request: ").append(urlPrefix).append(endpoint)
```
`System.out` instead of `System.err`

shivam-maharshi replied:
@risdenk these are not error messages but just information logs. That's why I didn't update them since they will carry the messages of all the HTTP codes including success like 200, 204 etc.

shivam-maharshi replied:
@risdenk if that's the norm we follow in YCSB then lemme make this change real quick and notify you back.

risdenk commented on rest/src/main/java/com/yahoo/ycsb/webservice/rest/RestClient.java:
```diff
@@ -0,0 +1,370 @@
+      responseCode = handleExceptions(e, urlPrefix + endpoint, HttpMethod.PUT);
+    }
+    if (logEnabled) {
+      System.out.println(new StringBuilder("PUT Request: ").append(urlPrefix).append(endpoint)
```
`System.out` instead of `System.err`

shivam-maharshi commented:
@risdenk I have updated all System.out to System.err. On of the three builds failed due to "build termination". Please take a look and lemme know. Thanks!

risdenk commented:
@shivam-maharshi looks good to me. Thanks!
